File: src/main/scala/io/arabesque/compression/SimpleStorage.scala
package io.arabesque.compression

import org.apache.hadoop.io.Writable
import java.io._
import java.util.concurrent.ExecutorService

import io.arabesque.computation.Computation
import io.arabesque.embedding.Embedding
import io.arabesque.odag.domain.{StorageReader, StorageStats}
import io.arabesque.pattern.Pattern
import io.arabesque.report.StorageReport
import io.arabesque.utils.Logging

/**
  * Created by ehussein on 6/28/17.
  */
abstract class SimpleStorage extends Writable with java.io.Externalizable with Logging {
  protected var storage:SimpleDomainStorage = _
  protected var serializeAsReadOnly: Boolean = false

  protected def createDomainStorage(readOnly: Boolean): SimpleDomainStorage =
    if (readOnly)
      new SimpleDomainStorageReadOnly
    else
      new SimpleDomainStorage

  def addEmbedding(embedding: Embedding): Unit

  def getPattern: Pattern

  def getReader(computation: Computation[Embedding], numPartitions: Int, numBlocks: Int, maxBlockSize: Int): StorageReader

  def aggregate(embZip: SimpleStorage): Unit

  def getNumberOfDomains(): Int = storage.getNumberOfDomains

  def getStorage: SimpleDomainStorage = storage

  def getNumberOfEnumerations: Long = storage.getNumberOfEnumerations

  def getNumberOfEmbeddings: Long = storage.getNumberOfEmbeddings

  def finalizeConstruction(pool: ExecutorService, numParts: Int): Unit = {
    storage.finalizeConstruction(pool, numParts)
  }

  def clear(): Unit = {
    storage.clear()
  }

  def getStats: StorageStats = storage.getStats

  def getSerializeAsWriteOnly: Boolean = serializeAsReadOnly

  def setSerializeAsReadOnly(serializeAsReadOnly: Boolean): Unit = {
    this.serializeAsReadOnly = serializeAsReadOnly
  }

  @throws[IOException]
  def writeInParts(outputs: Array[DataOutput], hasContent: Array[Boolean]): Unit = {
    storage.write(outputs, hasContent)
  }

  def printAllEnumerations(filePath: String) = storage.printAllEnumerations(filePath)

  def getStorageReport(): StorageReport = storage.getStorageReport()

  def getNumberSpuriousEmbeddings: Long = storage.getNumberSpuriousEmbeddings
}

File: src/main/java/io/arabesque/compression/SimpleStorage.java
package io.arabesque.compression;

import io.arabesque.computation.Computation;
import io.arabesque.embedding.Embedding;
import io.arabesque.odag.domain.StorageReader;
import io.arabesque.odag.domain.StorageStats;
import io.arabesque.pattern.Pattern;
import org.apache.hadoop.io.Writable;

import java.io.DataOutput;
import java.io.Externalizable;
import java.io.IOException;
import java.util.concurrent.ExecutorService;

public abstract class SimpleStorage implements Writable, Externalizable {
    protected UPSDomainStorage storage;
    protected boolean serializeAsReadOnly;

    protected UPSDomainStorage createDomainStorage(boolean readOnly) {
        if (readOnly) return new UPSDomainStorageReadOnly();
        else return new UPSDomainStorage();
    }

    public abstract void addEmbedding(Embedding embedding);
    public abstract Pattern getPattern();
    public abstract StorageReader getReader(
          Computation<Embedding> computation,
          int numPartitions,
          int numBlocks,
          int maxBlockSize);
    public abstract void aggregate(SimpleStorage embZip);

    public int getNumberOfDomains() {
        return storage.getNumberOfDomains();
    }

    public UPSDomainStorage getStorage() {
        return storage;
    }

    public long getNumberOfEnumerations() {
        return storage.getNumberOfEnumerations();
    }

    public void finalizeConstruction(ExecutorService pool, int numParts) {
        storage.finalizeConstruction(pool, numParts);
    }

    public void clear() {
        storage.clear();
    }

    public StorageStats getStats() {
        return storage.getStats();
    }

    public boolean getSerializeasWriteOnly() {
        return serializeAsReadOnly;
    }

    public void setSerializeAsReadOnly (boolean serializeAsReadOnly) {
        this.serializeAsReadOnly = serializeAsReadOnly;
    }

    public void writeInParts(DataOutput[] outputs, boolean[] hasContent) throws IOException {
        storage.write(outputs, hasContent);
    }

}
